shm/ship.hpp

```diff
@@ -0,0 +1,52 @@
+#pragma once
+#include <string>
+#include <vector>
+#include <memory>
+#include "cargo.hpp"
+
+class Ship {
+public:
+    Ship()
+        : id_(-1)
+    {}
+    Ship(int capacity, int maxCrew, int speed, const std::string& name, size_t id, std::vector<std::shared_ptr<Cargo>> cargo)
+        : capacity_(capacity)
+        , maxCrew_(maxCrew)
+        , crew_(0)
+        , speed_(speed)
+        , name_(name)
+        , id_(id)
+        , cargo_(cargo)
+    {}
+    Ship(int maxCrew, int speed, size_t id, std::vector<std::shared_ptr<Cargo>> cargo)
+        : Ship(0, maxCrew, speed, "", id, cargo)
+    {}
+
+    void setName(const std::string& name) { name_ = name; }
+
+    Ship& operator-=(size_t num) {
+        crew_ -= num;
+        return *this;
+    }
+    Ship& operator+=(size_t num) {
+        crew_ += num;
+        return *this;
+    }
```

Comment:
What happened when we reach the limit of capacity or drop below 0? You need to handle error cases. Maybe throw exception, maybe return `std::nullopt` or `nullptr` maybe set some specific value like `std::numeric_limits::min/max`.